### PR TITLE
Ruby 3.3 support

### PR DESCRIPTION
**Pull request by wildodecisiv**
### What does this PR do?

- Add suppport for ruby 3.3 generating the layers for amd and arm 64 architecture
- Remove `Gemfile.lock` from  `.gitignore` because it causes conflicts adding the local Gemfile.lock to the docker container trying to `bundle install`  dependencies
- Remove `-v` from `scritps/run_tests.sh` to avoid copying all the files from the local to the docker container and have in the container only the files specified in the `.dockerignore` 

### Motivation

At my current job we have many lambda functions running using ruby 3.2 and datadog layer, I am migrating to ruby 3.3 because it is [available since Apr 2, 2024](https://aws.amazon.com/es/about-aws/whats-new/2024/04/aws-lambda-ruby-3-3/).

In my team we are waiting this datadog updates for a couple of months but it seems you don't have a bandwith to work on this so far.

#90 

### Testing Guidelines

- Using the script `run_tests.sh`

### Additional Notes

**Would anyone of the main contributors want to `run the integrations tests` mentioned in the step 7th of the Contributing guide?**

Even having a valid DD_API_KEY I dont want to use my company's api key for running `integration tests` that can incur in additional charges  :smile: 
```
BUILD_LAYERS=true DD_API_KEY=<your Datadog api key> ./scripts/run_integration_tests.sh
```

### Types of changes

- [ ] Bug fix
- [x] New feature
- [ ] Breaking change
- [ ] Misc (docs, refactoring, dependency upgrade, etc.)

### Check all that apply

- [ ] This PR's description is comprehensive
- [ ] This PR contains breaking changes that are documented in the description
- [ ] This PR introduces new APIs or parameters that are documented and unlikely to change in the foreseeable future
- [x] This PR impacts documentation, and it has been updated (or a ticket has been logged)
- [ ] This PR's changes are covered by the automated tests
- [ ] This PR collects user input/sensitive content into Datadog

### CODE REVIEW

File: datadog-lambda.gemspec [hunk 1]
```diff
@@ -32,6 +32,14 @@ Gem::Specification.new do |spec|
...
   spec.add_dependency 'aws-xray-sdk', '~> 0.11.3'
   spec.add_dependency 'dogstatsd-ruby', '~> 5.0'
+
+  ruby_version = Gem::Version.new(RUBY_VERSION) # rubocop:disable Gemspec/RubyVersionGlobalsUsage
+  if ruby_version >= Gem::Version.new('3.3')
+    spec.add_dependency 'nokogiri', '~> 1.16.0'
+  else
+    spec.add_dependency 'nokogiri', '~> 1.15.6'
+  end
+
```
Comment (wildodecisiv) [inline datadog-lambda.gemspec:42]: Unfortunately nokogiri 1.15.6 does not work for ruby 3.3, so I had to explicitly set the nokogiri version when building the gem for ruby greather than or equal to 3.3.
Let me know if you find another way to handle this issue

File: scripts/Dockerfile_test [hunk 2]
```diff
@@ -1,8 +1,8 @@
 ARG image
 FROM $image
 
-COPY . datadog-lambda-ruby
-WORKDIR /datadog-lambda-ruby
+COPY . datadog-lambda-layer-ruby
```
Comment (wildodecisiv) [inline scripts/Dockerfile_test:4]: I renamed this to match the `-w` option definied in the docker run step of the `run_tests.sh`

File: scripts/run_tests.sh [hunk 3]
```diff
@@ -17,8 +17,7 @@ do
         -f scripts/Dockerfile_test . \
         --quiet \
         --build-arg image=ruby:$ruby_version
-    docker run --rm -v `pwd`:/datadog-lambda-layer-ruby \
-        -w /datadog-lambda-layer-ruby \
+    docker run --rm \
```
Comment (wildodecisiv) [inline scripts/run_tests.sh:20]: I think we no longer need the `-w` if the Dockerfile_tests already defined it

File: .dockerignore [hunk 4]
```diff
@@ -1,8 +1,8 @@
 *
+
```
Comment (wildodecisiv) [inline .dockerignore:2]: I removed Gemfile.lock from this to do not include it during docker building process, this causes issues when running bundle install over a Gemfile.lock copied from local to the container that was generating using a different ruby version

**Comment by wildodecisiv** [comment]
@duncanista  let me know if I need to do more changes on this pr please :)

File: lib/datadog/lambda/version.rb [hunk 5]
```diff
@@ -12,7 +12,7 @@ module Datadog
   module Lambda
     module VERSION
       MAJOR = 2
-      MINOR = 22
+      MINOR = 23
```
Comment (duncanista) [inline lib/datadog/lambda/version.rb:15]: Changing this to be its own PR

File: datadog-lambda.gemspec (hunk 1 above)
Comment (duncanista) [reply datadog-lambda.gemspec:42]: I think we can upgrade this, since I just deprecated Ruby 2.7 in another PR, let me test it

Comment (duncanista) [reply datadog-lambda.gemspec:42]: Yeah, this is still needed

**Review by duncanista — APPROVED** [review]

**Comment by duncanista** [comment]
Everything looks good!

Thanks for this amazing change 🙇🏽 
Will merge and start a release as soon as I can!

**Comment by wildodecisiv** [comment]
> Everything looks good!
> 
> Thanks for this amazing change 🙇🏽 Will merge and start a release as soon as I can!

That's great! thanks @duncanista !

File: datadog-lambda.gemspec (hunk 1 above)
Comment (cnorthwood) [reply datadog-lambda.gemspec:42]: there is an issue here, in that the Gem that was built and published onto Rubygems was built using 3.2, so the requirement in the Gemfile is set to `~> 1.15.6` (which has a CVE) - because this is in the gemspec even when installed on Ruby 3.3 it's set to this at compile time (see the dependencies list at

Comment (wildodecisiv) [reply datadog-lambda.gemspec:42]: good cath @cnorthwood! 

I think that is an issue due to the deployment step and not an issue with the conditional setting of the `nokogiri` gem based in the `RUBY_VERSION` variable.

Executing locally, if the `gem build` runs having `ruby v 3.3.0` it successfully installed the correct nokogiri version 

image

I am not sure how @duncanista deploys this but I am guessing is using the script `publish_prod.sh`

image

However, I think the published layer is ok, for ruby 3.3 is using the correct nokogiri version, this is because the layer is built into a docker container where RUBY_VERSION is set properly based on the docker image

image

I have the new layer working properly in a production project, this was not possible before this release.

Comment (duncanista) [reply datadog-lambda.gemspec:42]: I'll take a look tomorrow – thanks for bringing this out, I'll also make sure to have a hotfix on the Gem, I think indeed the layers are OK whereas the gem is probably using my local ruby, which I think I might have messed up – sorry.

Comment (duncanista) [reply datadog-lambda.gemspec:42]: Updating this in #97

Comment (duncanista) [reply datadog-lambda.gemspec:42]: Fixed in [v2.23.1](https://github.com/DataDog/datadog-lambda-rb/releases/tag/v2.23.1)
